libs/agent-sm/agent-policy/src/main/java/org/opensearch/secure_sm/policy/PolicyFile.java

```diff
@@ -0,0 +1,311 @@
+
+    private volatile PolicyInfo policyInfo;
+    private URL url;
+
+    /**
+     * When a policy file has a syntax error, the exception code may generate
+     * another permission check and this can cause the policy file to be parsed
+     * repeatedly, leading to a StackOverflowError or ClassCircularityError.
+     * To avoid this, this set is populated with policy files that have been
+     * previously parsed and have syntax errors, so that they can be
+     * subsequently ignored.
+     */
+    private static Set<URL> badPolicyURLs = Collections.newSetFromMap(new ConcurrentHashMap<URL, Boolean>());
+
+    /**
+     * Initializes the Policy object and reads the default policy
+     * from the specified URL only.
+     */
+    public PolicyFile(URL url) {
+        this.url = url;
+        init(url);
+    }
+
+    /**
+     * Initializes the Policy object and reads the default policy
+     * configuration file(s) into the Policy object.
+     *
+     * See the class description for details on the algorithm used to
+     * initialize the Policy object.
+     */
+    private void init(URL url) {
+        PolicyInfo newInfo = new PolicyInfo();
+        initPolicyInfof(url, newInfo);
+        policyInfo = newInfo;
+    }
+
+    private void initPolicyInfof(URL policy, PolicyInfo newInfo) {
+        if (badPolicyURLs.contains(policy)) {
+            return;
+        }
+
+        try (InputStreamReader reader = new InputStreamReader(getInputStream(policy), StandardCharsets.UTF_8)) {
+            PolicyParser policyParser = new PolicyParser();
+            policyParser.read(reader);
+
+            Collections.list(policyParser.grantElements()).forEach(grantNode -> {
+                try {
+                    addGrantNode(grantNode, newInfo);
+                } catch (Exception e) {
+                    e.printStackTrace(System.err);
+                }
+            });
+
+            return;
+        } catch (PolicyParser.ParsingException pe) {
+            badPolicyURLs.add(policy);
+            pe.printStackTrace(System.err);
+        } catch (IOException ioe) {
+            ioe.printStackTrace(System.err);
+        }
+
+        return;
+    }
+
+    /**
+     * Refreshes the policy object by re-reading all the policy files.
+     */
+    @Override
+    public void refresh() {
+        init(url);
+    }
+
+    @Override
+    public boolean implies(ProtectionDomain pd, Permission p) {
+        PermissionCollection pc = getPermissions(pd);
+        if (pc == null) {
+            return false;
+        }
+
+        // cache mapping of protection domain to its PermissionCollection
+        return pc.implies(p);
+    }
+
+    @Override
+    public PermissionCollection getPermissions(ProtectionDomain domain) {
+        Permissions perms = new Permissions();
+
+        if (domain == null) return perms;
+
+        getPermissionsForProtectionDomain(perms, domain);
+
+        // - adding static perms after policy perms is necessary
```

Comment:
What is a static perm? In what cases would the block below add more perms?

Reply:
removed the stale comment